mhkit/power/characteristics.py

```diff
@@ -1,121 +1,185 @@
+        Measured DC voltage [V] indexed by time
+
+    current: pandas Series, pandas DataFrame, xarray DataArray, or xarray Dataset
+        Measured three phase current [A] indexed by time
+
     power_factor: float 
         Power factor for the efficiency of the system
-    line_to_line: bool
+
+    line_to_line: bool (Optional)
         Set to true if the given voltage measurements are line_to_line
-    
+
+    to_pandas: bool (Optional)
+        Flag to save output to pandas instead of xarray. Default = True.
+
     Returns
     --------
-    P: pandas DataFrame
+    P: pandas DataFrame or xarray Dataset
         Magnitude of active AC power [W] indexed by time with Power column 
     """
-    if not isinstance(voltage, pd.DataFrame):
-        raise TypeError(f'voltage must be of type pd.DataFrame. Got: {type(voltage)}')
-    if not isinstance(current, pd.DataFrame):
-        raise TypeError(f'current must be of type pd.DataFrame. Got: {type(current)}')
-    if not len(voltage.columns) == 3:
+    if not isinstance(voltage, (pd.Series, pd.DataFrame, xr.DataArray, xr.Dataset)):
+        raise TypeError('voltage must be of type pd.Series, pd.DataFrame, ' + 
+                        f'xr.DataArray, or xr.Dataset. Got{type(voltage)}')
+    if not isinstance(current, (pd.Series, pd.DataFrame, xr.DataArray, xr.Dataset)):
+        raise TypeError('current must be of type pd.Series, pd.DataFrame, ' + 
+                        f'xr.DataArray, or xr.Dataset. Got{type(current)}')
+
+    # Convert inputs to xr.Dataset
+    voltage = _convert_to_dataset(voltage, 'voltage')
+    current = _convert_to_dataset(current, 'current')
+
+    # Check that sizes are the same
+    if not len(voltage.data_vars) == 3:
         raise ValueError('voltage must have three columns')
-    if not len(current.columns) == 3:
+    if not len(current.data_vars) == 3:
         raise ValueError('current must have three columns')
-    if not current.shape == voltage.shape:
+    if not current.sizes == voltage.sizes:
         raise ValueError('current and voltage must be of the same size')
-    
 
-    abs_current = np.abs(current.values)
-    abs_voltage = np.abs(voltage.values)
+    power = dc_power(voltage, current, to_pandas=False)['Gross']
+    power.name = 'Power'
+    power = power.to_dataset() # force xr.DataArray to be consistently in xr.Dataset format
+    P = np.abs(power) * power_factor
 
     if line_to_line:
-        power = abs_current * (abs_voltage * np.sqrt(3))
-    else:
-        power = abs_current * abs_voltage
-         
-    power = pd.DataFrame(power) 
-    P = power.sum(axis=1) * power_factor
-    P = P.to_frame('Power')
-    
+        P = P * np.sqrt(3)
+
+    if to_pandas:
+        P = P.to_pandas()
+
     return P
+
+def _convert_to_dataset(data, name='data'):
+    # Takes data that could be pd.DataFrame, pd.Series, xr.DataArray, or 
+    # xr.Dataset and converts it to xr.Dataset
+    if isinstance(data, (pd.DataFrame, pd.Series)):
```

Comment:
Let's give this a formal docstring and type checking for production.

    """
    Converts the given data to an xarray.Dataset.

    This function is designed to handle inputs that can be either a pandas DataFrame, a pandas Series,
    an xarray DataArray, or an xarray Dataset. It ensures that the output is consistently an xarray.Dataset.

    Parameters
    ----------
    data: pandas DataFrame, pandas Series, xarray DataArray, or xarray Dataset
        The data to be converted. 

    name: str (Optional)
        The name to assign to the data variable in case the input is an xarray DataArray without a name.
        Default value is 'data'.

    Returns
    -------
    xarray.Dataset
        The input data converted to an xarray.Dataset. If the input is already an xarray.Dataset,
        it is returned as is.

    Examples
    --------
    >>> df = pd.DataFrame({'A': [1, 2, 3], 'B': [4, 5, 6]})
    >>> ds = _convert_to_dataset(df)
    >>> type(ds)
    <class 'xarray.core.dataset.Dataset'>

    >>> series = pd.Series([1, 2, 3], name='C')
    >>> ds = _convert_to_dataset(series)
    >>> type(ds)
    <class 'xarray.core.dataset.Dataset'>

    >>> data_array = xr.DataArray([1, 2, 3])
    >>> ds = _convert_to_dataset(data_array, name='D')
    >>> type(ds)
    <class 'xarray.core.dataset.Dataset'>
    """
    if not isinstance(data, (pd.DataFrame, pd.Series, xr.DataArray, xr.Dataset)):
        raise TypeError("Input data must be of type pandas.DataFrame, pandas.Series, "
                    "xarray.DataArray, or xarray.Dataset")
                        
    if not isinstance(name, str):
        raise TypeError("The 'name' parameter must be a string")

Reply:
Agreed, done. Might have to move this to a more general location later on